tests/components/sensor/test_awair.py

```diff
@@ -0,0 +1,249 @@
+
+    if config is None:
+        config = DISCOVERY_CONFIG
+
+    with devices_patch, air_data_patch:
+        assert await async_setup_component(hass, SENSOR_DOMAIN, config)
+
+    await hass.async_block_till_done()
+
+
+async def test_platform_manually_configured(hass):
+    """Test that we can manually configure devices."""
+    await setup_awair(hass, MANUAL_CONFIG)
+
+    assert len(hass.states.async_all()) == 6
+
+    # Ensure that we loaded the device with uuid 'awair_foo', not the
+    # 'awair_12345' device that we stub out for API device discovery
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_co2')
+    assert entity.unique_id == 'awair_foo_CO2'
+
+
+async def test_platform_automatically_configured(hass):
+    """Test that we can discover devices from the API."""
+    await setup_awair(hass)
+
+    assert len(hass.states.async_all()) == 6
+
+    # Ensure that we loaded the device with uuid 'awair_12345', which is
+    # the device that we stub out for API device discovery
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_co2')
+    assert entity.unique_id == 'awair_12345_CO2'
+
+
+async def test_bad_platform_setup(hass):
+    """Tests that we throw correct exceptions when setting up Awair."""
+    from python_awair import AwairClient
+
+    auth_patch = patch('python_awair.AwairClient.devices',
+                       side_effect=AwairClient.AuthError)
+    rate_patch = patch('python_awair.AwairClient.devices',
+                       side_effect=AwairClient.RatelimitError)
+    generic_patch = patch('python_awair.AwairClient.devices',
+                          side_effect=AwairClient.GenericError)
+
+    with auth_patch:
+        assert await async_setup_component(hass, SENSOR_DOMAIN,
+                                           DISCOVERY_CONFIG)
+        assert not hass.states.async_all()
+
+    with rate_patch:
+        assert await async_setup_component(hass, SENSOR_DOMAIN,
+                                           DISCOVERY_CONFIG)
+        assert not hass.states.async_all()
+
+    with generic_patch:
+        assert await async_setup_component(hass, SENSOR_DOMAIN,
+                                           DISCOVERY_CONFIG)
+        assert not hass.states.async_all()
+
+
+async def test_awair_attributes(hass):
+    """Test that desired attributes are set."""
+    await setup_awair(hass)
+
+    attributes = hass.states.get('sensor.awair_co2').attributes
+    fixture = json.loads(load_fixture('awair_air_data_latest.json'))
+    timestamp = datetime.strptime(fixture[0]['timestamp'], TIME_FORMAT)
+    assert attributes['timestamp'] == timestamp
+
+
+async def test_awair_score(hass):
+    """Test that we create a sensor for the 'Awair score'."""
+    await setup_awair(hass)
+
+    sensor = hass.states.get('sensor.awair_score')
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_score')
+    assert sensor.state == '78'
+    assert entity.device_class == DEVICE_CLASS_SCORE
```

Comment:
These entity properties are also state attributes. Assert the state attributes instead. We want to avoid asserting things about the entities, since that makes the tests less robust.

Reply:
> These entity properties are also state attributes.
I was getting ready to quote the [documentation](https://dev-docs.home-assistant.io/en/master/api/core.html#homeassistant.core.State), but I played around with a debugger for a bit and realized that they're ... _attributes_. Not of the state object itself (which is how I interpreted your comment at first), but like `state.attributes['device_class']` attributes. :smile: 

I'll fix these - thanks. I was kinda sad about having to drop down into the entity!